### DISCUSSION TITLE
Using go client to access  a Harbor registry

### DISCUSSION THREAD

**Discussion by dougchristie-DukeEnergy** [moby/moby#45360] [General]
I am have been seeing odd behavior when using the the go client to access a harbor registry. On smy windows machine There appears to be a race condition. If I slow things thing they appear to work.All calls to RegistryLogin, ImageTag and ImagePush all return no errors, but the image doesn't show up in the harbor registry. When I try this on an AWS workspace running ubuntu. Everything returns 200 when digging into the API call but nothing shows up in the registry. When  I tried to pull an image from harbor I get now error and see a 200 digging into the api, but no image appears in the local registry when I run "docker image ls". Any idea why this happens.

**Comment by thaJeztah** [comment]
The push and pull actions provide a streaming response; the 200 status code is returned when the request is received, but a streaming response containing the push/pull progress (and any errors occurring during push or pull) will follow. So when you receive the 200 status code, the push/pull is likely still in progress; also see the docker/cli code as an example https://github.com/docker/cli/blob/26a73575f1e3a0d1dc6ee6592db5926dcb907feb/cli/command/image/trust.go#L285

**Reply by dougchristie-DukeEnergy** [reply]
So when I make a call to func (*Client) ImagePush (https://pkg.go.dev/github.com/docker/docker/client?utm_source=godoc#Client.ImagePush) How do I check the stream to see if the image push was successful?

**Reply by thaJeztah** [reply]
Basically it comes down to continue reading the stream (there's the `DisplayJSONMessagesToStream ` utility that the CLI uses); https://github.com/moby/moby/blob/f117aef2ea63ee008c05a7506c8c9c50a1fa0c7f/pkg/jsonmessage/jsonmessage.go#L285-L289

That function is to also print the output (which may not be needed strictly for your case), but it iterates over the stream returned, and decodes each progress message; https://github.com/moby/moby/blob/f117aef2ea63ee008c05a7506c8c9c50a1fa0c7f/pkg/jsonmessage/jsonmessage.go#L216-L224

As part of looping over those messages, there's the `Display()` method; https://github.com/moby/moby/blob/f117aef2ea63ee008c05a7506c8c9c50a1fa0c7f/pkg/jsonmessage/jsonmessage.go#L263

Which checks if the progress message has an Error set; that Error will contain details if something failed as part of the push/pull
https://github.com/moby/moby/blob/f117aef2ea63ee008c05a7506c8c9c50a1fa0c7f/pkg/jsonmessage/jsonmessage.go#L169-L175


It's definitely not the most elegant part of our code (and we should consider having utilities to handle that stream separate from printing it, or have a function on the client that handles more of this, but that's not there (yet) 😅 

Hope this helps!

**Reply by dougchristie-DukeEnergy** [reply]
It definitely clears up some things.  Only thing is if I make the following call "ioReader, err := dockerClient.ImagePush(ctx, image, options)" all that is returned is an ioReader and err. If I want to call  DisplayJSONMessagesToStream where does the stream come from, what would the call back function look like.  I assume I use the ioReader return as the first argument?

**Reply by dougchristie-DukeEnergy** [reply]
Anyway I can get an answer to my question above?

**Reply by dougchristie-DukeEnergy** [reply]
I really need to know how to call and check the stream. We are trying to build a product based on these libraries and it would be really help to know how to check that stream.
